### PR TITLE
resource blacklist API

### PR DESCRIPTION
**Pull request by opqdonut**
for #1682

# Definition of Done / Review checklist

## Reviewability
- [x] link to issue

## API
- [x] API is documented and shows up in Swagger UI
- [x] API is backwards compatible or completely new
- [x] Events are backwards compatible

## Documentation
- [x] update changelog if necessary (will update once feature is complete)

## Testing
- [x] complex logic is unit tested
- [x] valuable features are integration / browser / acceptance tested automatically

## Follow-up
- [ ] no critical TODOs left to implement

### CODE REVIEW

File: resources/migrations/20191007092006-blacklist.up.sql [hunk 1]
```diff
@@ -0,0 +1,8 @@
+CREATE TABLE blacklist_event (
+  id serial NOT NULL PRIMARY KEY,
+  eventdata jsonb
+);
+--;;
+CREATE INDEX index_blacklist_event_resource ON blacklist_event((eventdata->>'blacklist/resource'));
+--;;
+CREATE INDEX index_blacklist_event_user ON blacklist_event((eventdata->>'blacklist/user'));
```
Comment (luontola) [inline resources/migrations/20191007092006-blacklist.up.sql]: To serve queries where both resource and user is used in where,  an index which contains both of the fields would serve it better.

Comment (luontola) [reply resources/migrations/20191007092006-blacklist.up.sql]: e.g. one index for (resource, user) and anther index for (user)

Comment (opqdonut) [reply resources/migrations/20191007092006-blacklist.up.sql]: good point. I'll start with a (resource, user) index

File: src/clj/rems/api/blacklist.clj [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    :tags ["blacklist"]
+    (GET "/" []
+      :summary "Get blacklist entries"
+      :roles #{:handler :owner}
+      :query-params [{user :- blacklist/UserId nil}
+                     {resource :- blacklist/ResourceId nil}]
+      :return BlacklistResponse
+      (ok (blacklist/get-blacklist {:blacklist/user user
+                                    :blacklist/resource resource})))
+    (POST "/command" []
```
Comment (luontola) [inline src/clj/rems/api/blacklist.clj]: /add and /remove

File: src/clj/rems/db/blacklist.clj [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+(defn- event->json [event]
+  (s/validate BlacklistEvent event)
+  (json/generate-string event))
+
+(defn add-event! [event]
+  (db/add-blacklist-event! {:eventdata (event->json event)}))
+
+(defn- event-from-db [event]
+  (assoc (json->event (:eventdata event))
+         :event/id (:id event))) ;; TODO :event/id ranges conflict between blacklist_event and application_event
```
Comment (luontola) [inline src/clj/rems/db/blacklist.clj]: Should not be a problem.

File: src/clj/rems/api/blacklist.clj [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  (:require [clj-time.core :as time]
+            [compojure.api.sweet :refer :all]
+            [rems.api.schema :as schema]
+            [rems.db.blacklist :as blacklist]
+            [rems.util :refer [getx-user-id]]
+            [ring.util.http-response :refer [ok]]
+            [schema.core :as s]))
+
+(s/defschema BlacklistCommand
+  ;; TODO should these be namespaced keywords?
```
Comment (luontola) [inline src/clj/rems/api/blacklist.clj]: The other command APIs don't use namespaced keywords, so it's okay the way it's now. We can add them later everywhere if needed.

File: src/clj/rems/api/blacklist.clj (hunk 2 above)
Comment (opqdonut) [reply src/clj/rems/api/blacklist.clj]: done

**Pull request merged**